### DISCUSSION TITLE
Basic Note Templates

### DISCUSSION THREAD

**Discussion by satchelfrench** [athensresearch/athens#875] [Feature Requests]
Please search our GitHub and Discord to see if your feature has already been requested!

**Why?**
- Templates are a small but _super_ powerful feature in note taking apps, especially daily notes.
- The feature is easy enough to implement, and would make being effective in Athens easier.
- Should other features like SmartBlocks be implemented, templates can always be merged with it.

**How?**
- The approach that roam takes is my personal favourite, but I think even simpler solutions would work also. For example a hot key 'Command + T' / button to bring up a templates list and select from -> then copy/pasting into current page (similar to Alfred)

**Comment by juniusfree** [comment]
Just to add on the why. Having a template reduces the effort of the users in capturing their thoughts (e.g. "fear of blank the page").

What other ways can Athens do to help the users capture their thoughts other than templates?
- Daily Notes
- https://github.com/athensresearch/athens/discussions/888

Benefits for Athens:
- The more the users capture their thoughts in Athens, the higher the switching cost and the more Athens can help in the latter part of the process (e.g. more data to analyze in making a conclusion)

Other related links:
- https://subconscious.substack.com/p/unconscious-r-and-d
- https://twitter.com/Conaw/status/1263885258618834945
- https://www.cortexfutura.com/getting-started-algorithms-of-thought/

More examples of templates:
- https://www.intelligentchange.com/products/the-five-minute-journal
- https://woopmylife.org/en/home

**Comment by solarkraft** [comment]
>  The approach that roam takes is my personal favourite, but I think even simpler solutions would work also. For example a hot key 'Command + T' / button to bring up a templates list and select from -> then copy/pasting into current page (similar to Alfred)

The Athens approach could be `/template <template name>`, perhaps with special template names that are automatically applied (daily notes are also my primary use case). 

I currently keep my daily template in a page called `template: daily` that is bookmarked in the side bar. Another use case would be one for meetings. It'd be cool if one could directly choose the template while creating the page (but I'm not sure how this would best be realized UX-wise yet).

**Reply by jefftangx** [reply]
I always wanted templates to be directly within the `/` menu. Typing `/template` is kind a lot! What do you think?

**Reply by solarkraft** [reply]
It's a cool idea since it'd make them first-class citizens. As long as there aren't too many templates (I can only think of a few for my use) the clutter and name conflict risks I see could stay negligible. 
I like it!

Thing about naming: Since  I see / as an "insert" command ("insert current time", "insert block embed" - "template: daily" would be too much to type again and "insert daily" sounds weird, doesn't it?) perhaps the template name could be rewritten to "daily template".

**Reply by shanberg** [reply]
How about surfacing "activate template <daily/meeting/notes>" as slash menu options, so you can type "template" to filter down to all templates, or "daily" to filter to just the daily template?

Assuming we'd do some more work on the terminology, because "activate" sounds silly.

**Reply by jefftangx** [reply]
How would templates be created and edited?

**LogSeq**: [LogSeq Templates](https://twitter.com/logseq/status/1308763162779373569?lang=en) are persisted by adding meta-data at the block-level via [org-mode property syntax](https://orgmode.org/guide/Properties.html), e.g.:

```
:PROPERTIES:
:heading: true
:template: template1
:END:
```

LogSeq templates are triggered by using a slash command `/template` which then creates a secondary dropdown for all the templates.

**Roam**: [Roam Templates](https://nesslabs.com/roam-research-templates-tutorial) use the special `roam/templates` page. They are triggered by pressing `;;`

Similarities between LogSeq and Roam: templates are created at the block-level, from anywhere.

Differences: LogSeq follows org-mode pattern, per usual. Roam follows namespace pattern, per usual (there's also roam/js and roam/css).

Can we do better? Longer-term: what about parameterizing? I use [espanso](https://espanso.org/) because it's cross-platform, but heard good things about Keyboard Maestro and TextExpander as well.

Side note: what if templates were created and stored in settings? I get why it makes sense to "Make Template" from any block, from anywhere. But I feel like the meta-data of a template might get too complex to be rendered by plain-text, without going deeper down making a custom language. This perhaps applies to tables in the future, or other more sophisticated components like kanban boards that do not naturally conform to outliner/block-based hierarchy.

**Reply by shanberg** [reply]
Creating templates outside the normal block edit flow probably allows for a clearer UX, because you can specialize that interface for creating a template by providing hints, dropdowns, text fields, etc.

Creating templates inside the normal block edit flow probably creates a more seamless experience, more amenable to shortcuts and power-user interactions.

A middle ground might be to make sure the template code is all directly editable, but provide very clear autocomplete and validation so you can't easily break your own templates, and also provide plenty of hints about what templates can do, but that's probably the most difficult option.

**Reply by juniusfree** [reply]
Adding here for reference. Here's how superhuman create/trigger a snippet:

- Create snippet
![image](https://user-images.githubusercontent.com/8164667/115498891-f099a400-a2a0-11eb-80aa-a17bdb920b04.png)

- Trigger snippet (`Ctrl+;`)
![image](https://user-images.githubusercontent.com/8164667/115498900-f4c5c180-a2a0-11eb-811d-f60a07dff3dc.png)

**Reply by jefftangx** [reply]
@juniusfree do you use superhuman??

**Reply by juniusfree** [reply]
@tangjeff0 nope. I'm just looking for some examples when found that.

**Reply by lukaskawerau** [reply]
(<- [CortexFutura](https://twitter.com/cortexfutura))

Some considerations from using templates daily, extensively, in Roam:

I prefer the `;;` shortcut to using `/` because it doesn't mix "domains" if that makes sense. If I type the shortcut, I'm immediately in a search interface for all my templates, without any further keystrokes to get to a sub-menu and the menu isn't filled with lots of options any time I'm typing `/` to use something other than a template.

I'm strongly in favor of parameterizing things and capabilities á la [SmartBlocks](https://github.com/athensresearch/athens/issues/568), it's one of my main annoyances in Roam that I have to install an extension to get this because it's not native. I understand that Roam42 has basically developed a new language for the logic, but having its functionality native in the app would be incredible. 

With regards to how templates are implemented, I can only say that I find Roam's namespacing approach to be fine – I have no experience with Logseq's approach. That said, I'm hugely in favor of any approach that allows me to skip using the mouse at all. It seems Logseq might require that, but that's only judging by their intro video.

**Reply by satchelfrench** [reply]
> I always wanted templates to be directly within the `/` menu. Typing `/template` is kind a lot! What do you think?

Actually after using this in logseq, it's become my personal favourite way to access templates!

**Reply by jefftangx** [reply]
@lukaskawerau I'm not a huge fan of `;;` to access SmartBlocks. This is another shortcut that users have to learn.

**Reply by mikejones3** [reply]
I honestly like how templates are broken out from the / commands in Roam. I like how its ;;templateName 

I think its quicker to find a template that way so you are not competing with the / commands.

**Comment by shanberg** [comment]
Another example: https://community.coda.io/t/launched-custom-templates-drag-drop-and-slash-your-way/22718

**Comment by Ejipuh** [comment]
What if templates were interactive--something like forms or a command line interface?

**Reply by jefftangx** [reply]
Can you provide a demo or illustration or what this would look like?
